### ISSUE TITLE
Unable to launch slack-machine

### ISSUE THREAD

**Issue by thatside** [DandyDev/slack-machine#1]
When I try to launch the app, an error comes in: 
```
Traceback (most recent call last):
  File "/home/thatside/own/venv/linker/bin/slack-machine", line 11, in <module>
    sys.exit(main())
  File "/home/thatside/own/venv/linker/lib/python3.6/site-packages/machine/bin/run.py", line 13, in main
    bot.run()
  File "/home/thatside/own/venv/linker/lib/python3.6/site-packages/machine/core.py", line 96, in run
    self._dispatcher.start()
  File "/home/thatside/own/venv/linker/lib/python3.6/site-packages/machine/dispatch.py", line 21, in start
    for event in self._client.rtm_read():
  File "/home/thatside/own/venv/linker/lib/python3.6/site-packages/slackclient/client.py", line 128, in rtm_read
    json_data = self.server.websocket_safe_read()
  File "/home/thatside/own/venv/linker/lib/python3.6/site-packages/slackclient/server.py", line 186, in websocket_safe_read
    data += "{0}\n".format(self.websocket.recv())
AttributeError: 'NoneType' object has no attribute 'recv'
```

Are there some nuances related to Python versions compatibility?

**Comment by thatside** [comment]
Sorry, seems there is another problem - this package is currently incompatible with Slack App-integrated bots. 
Was too impatient to try the package :)
Closing the issue now. 
Hope it will support Slack apps soon - AFAIK Slack team recommends this way.

**Comment by DandyDev** [comment]
Hi! I want to support your use case, but I'm not sure exactly what you want/need.
How do "Slack App-integrated bots" differ from bots you register through https://my.slack.com/services/new/bot ?

**Comment by thatside** [comment]
Hello!
As stated here: https://api.slack.com/bot-users: every workspace (team) could have custom integration among which bot users go. But bots which could be used by other teams (theoritically :)) should be attached to Slack app itself and then they could be used by all teams who have the app installed. 
So you should create an app here first: https://api.slack.com/apps. And then you could register a bot user for this app.
This isn't critical for me - I'm developing a simple app for my team but it would be more convenient in future to be able to use both token types (AFAIK simple bot token starts with `xoxb` and app token starts with `xoxp`).

I've found the package on pypi using search - it has high "search weight" so I wanted to find an alternative of using simple web server with slash commands for interaction :)

**Comment by DandyDev** [comment]
Interesting concept. I had heard about Slack apps, and even installed a couple for my own teams. But I've not looked closely into how Slack apps work from the technical perspective.

> But bots which could be used by other teams (theoritically :)) should be attached to Slack app itself and then they could be used by all teams who have the app installed.

This technically means that the bot should be able to handle multiple teams, right? This is currently not the case for Slack Machine. It uses Slack's own Python Slack client, which - as far as I can tell - assumes it is connected to 1 team only.

**Comment by thatside** [comment]
I've checked slackclient package and it doesn't seems it could handle only one team at the given moment - neither it isn't said it could handle multiple teams. Should check this in future.

Nevertheless my initial problem was solved and I think it is more important to develop other features than add support of multi-team applications :)

**Comment by DandyDev** [comment]
Agreed. Multi-team/App support are currently out of scope. I did find an example from Slack itself on how to build a Slack App in Python: https://github.com/slackapi/Slack-Python-Onboarding-Tutorial. But that example [doesn't seem to support multiple teams](https://github.com/slackapi/Slack-Python-Onboarding-Tutorial/issues/5).

If Slack Machine were to support being installed as a Slack App, it would have to implement the OAuth flow, and with that, it would have to run a web server (like Flask) to be accessible from the outside. I actually like the fact that currently, it doesn't need to be accessible from the outside (except for the websocket connection of course), so it can be run behind a firewall and doesn't have to expose a web server.

I'll leave it alone for now :)

If I may ask though, how did you solve your issue?

**Comment by DandyDev** [comment]
Update: what's interesting, is that Hubot also doesn't support multiple teams as far as I can tell. And when you install the Hubot Slack App, it will give you an `xoxb` token:

![](https://dandydev.net/share/Screen-Shot-2017-09-18-11-31-09.png) (don't worry, the integration was removed, so the token is invalid)

So apparently, you can create Slack apps that do not use OAuth?

**Comment by thatside** [comment]
> If I may ask though, how did you solve your issue?
I have created a separate bot user - first I have used the token from Slack App and it doesn't work but with separate bot token (from the link you have provided) everything worked perfectly.

Hubot seems to be a slightly different thing - it seems it only provides an adapter to generate a bot token in more convenient way. 

At least it is needed to check if it is possible to use app bot tokens (and how to implement this if not possible) - for example, if somebody would try to implement OAuth part on their side and then start the machine with tokens received from team admin.
I'll check this out when I'll have my free time :)
